
# Homepage HTML in string format
acronym_string = '''Novel Extreme-resolution Metabolomics Streamlined data analysis Integrated Software'''

homepage_string = '''This software is geared towards FT-ICR-MS extreme-resolution <strong>mass spectrometry data</strong>.
<br>
<br>
- It is suited to read direct-infusion mass spectrometry data. It can however read any
other type of data if it follows the formatting rules as declared in the "Data Reading" page.
<br>
- Thus, it can handle not only Direct Infusion FT-ICR-MS data but also LC-MS and GC-MS data. It will not use the time
retention parameters of the latter 2 and as such may not be the most suitable tool to analyse said data.
<br>
<br>
- This software includes reading the data, data filtering, data annotation based on mass (or <em>m/z</em> peak)
deviation, data pre-treatment, univariate, multivariate unsupervised and supervised analysis, common and exclusive
compound analysis, data diversity visualization tools (Van Krevelen, Kendrick Mass Defect Plots and Chemical Composition
Series), Pathway matching between identified HMDB based IDs and Over-Representation analysis.
<br>
- It also includes statistical analysis performe on data treated with Binary Simplification (BinSim),
that is, feature occurrence data that can be useful for extreme-resolution data with plenty of missing values.
<br>
- Data Annotation is made based on the metabolite databases given by the user. This database should be formatted according
to the indications in the "Data Annotation" page.<br>
- Every Figure generated by the software can be downloaded (for non-interactive figures a button will be pressable to
save them, and for interactive figures a symbol on their upper right can be used to save them) and a final report of
the analysis can also be generated.
<br>
<br>
<hr />
<br>
- Software made by: <strong>FT-ICR and Structural Mass Spectrometry Laboratory, BioISI, Faculdade de CiÃªncias da
Universidade de Lisboa</strong> (<a href="http://ft-icr.rd.ciencias.ulisboa.pt/" target="_blank"
rel="nofollow">ft-icr.rd.ciencias.ulisboa.pt/</a>)
<br>
<br>
If you use this software in your untargeted metabolomcis data analysis, we would be grateful if you cite our paper
introducing it:
<br>
<br>
- Paper Link: <strong>(LINK)</strong><br>
- Citation: <strong>(CITATION)</strong>
'''


# Instruction page HTML in string format

instructions_page = '''<p>Here, we will go through some general considerations to have while using this software by describing
how it works and some of the currently known issues overall that exist. Furthermore, if you encounter an issue or problem
with the software feel free to inform us at
<a data-fr-linked="true" href="https://github.com/fticr-sms/Metabolomics_DataAnalysis_Pipeline/issues/new">https://github.com/fticr-sms/Metabolomics_DataAnalysis_Pipeline/issues/new</a>.
If you feel some parts are underexplained or cannot be intuitively grasped, please also inform us so we can improve the
descriptions provided.</p>
<p><strong>General Considerations to Have</strong></p>
<p>- If you have reached this page, we are assuming that you have all the needed required Python packages and their versions
that were outlined in the README.md to run this software.</p>
<p>- This software is both presented in Jupyter Notebook format as well as this graphical interface. Each of them has unique
parts and sections that are only present in one of these formats. In general the former can have results and outputs be
more easily editable and customized to the user (requiring basic programming background) while the latter is more suited
for a quick analysis for users less experienced with Python programming at the cost of not being able to expand the
analysis further than what is already present in the software.</p>
<p>- While on data pre-processing and pre-treatment, steps will be unlocked one by one starting with <strong>Data
Reading</strong> which is the only one unlocked at the begginning. When reaching the final step of this section in
<strong>Class Colours</strong>, after selecting <u>Next Step - Analysis</u>, <strong>all analysis steps</strong> become
unlocked (as well as <strong>Report Generation</strong>). These steps are generally
independent of one another and can be done in any order or skipped based on the type of analysis to be performed. As with
every Metabolomics experiment, statistical analysis results will depend on the filtering and pre-treatment methods used.</p>
<p>- If you <strong>change any part of the pre-treatment steps</strong>, posterior pre-treatment steps and statistical
analysis will immediately <strong>be locked</strong> and will <strong>erase previous statistical analysis</strong>, so you
will have to repeat data pre-treatments. This is to avoid situations where the user wanted to change the pre-treatment but
forgot to re-apply posterior pre-treatment steps.</p>
<p>- While there is an effort for figures and tables produced that can be downloaded to have the relevant parameters used
to build them in the name, it is a good practice to keep in mind what parameters were used to make these figures and tables,
especially those that do not update with every parameter change to safeguard possible errors that may arise in the filenames
generated.</p>
<p>- All tables and figures are saved in the 'Downloads' folder. Interactive figures are downloaded using the toolbar that appears at the
top right of the figure when hovering over it. Non-interactive figures are saved using an available button in the software. The
generated report folder is put on the current working directory, that is, the folder where NEMeSIS is.</p>
<p>- When you start running something there is not a way to stop the process. You can observe if the program is running by
observing the little circle on the top right. If there is something currently running it will have this aspect:
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAtCAYAAADsvzj/AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAFxEAABcRAcom8z8AAAl7SURBVGhD3Vh5WNTXFUVFjdFEm5oYTe2X5jPWvSqCyqLgVhuS1lh3bZtqTGzQKIiKxqVapFqNSzQ22hqTVE1iSqKILMpiBFEQlwgIuLCJ7MvMIGpM0tNz3zDpdPpwGPAf+sf5ZrjM771737n33Pt+Tk4Bx/B/Aa2xOUJrbI7QGpsjtMbmCK2xsfAPh9Mi4s2j/DsKTktj4bQkxgYn4BQYDafF/P/iyP9do7HQGh2BPyHOyyf/brs0Et3WxqJr4KfoMHsX2s/dg8de3Y0Oc/9G7MGj8/bBecFnaLXwc7TwZ8CWoGzXdRRaY0MhzqtAjuGJlccxdFsCXt6bDL8DSfiVfwh6jnwZPx03A73HTkOvn88iZqCH7xx0nbISnWdtQMc5O9HWb79VQE1gSGtsCOqC6LQiGs+vj8OMD8/hYEoezlwrQnj8acz1W4T+g4dj8LARGOI+kvCGi7sP4Y1BXuPRZ8wUFdyzE+aj0+ztaLnoCAOJ4NoCm70aAq3RHhQL4WgfFMUAUrH39DUkZt9CWaUBtbdNKCq8iXXrgtF/oCuGuLnDbZgnXId5Kch3gQrMYxQGe41F37FT8KNJy9B+3vtkh2naGGa0xgdhYTha8uSe/VMsJr2fgojLBag2GFBjMuBOjQnf3L0N4D4+/SwU7r5T0MdnAvqMnqScdvEcjUEjxsNtqAeD8VJMWYL7mbcvuk8MUHXUyj+sjh3N/vVBa3wAWpCN50PisfRwGqLTClBeWY37tTX45k4NviVyiysRkXYLr++NQ4952/HErI34wW83o/uvl+CZycvRdepK9Bk7Fa7DR2LoUHd+/if1Bni/yN8F4pE39ptrxpE00xrrw4IwdKciLTuShrPXi8lENW6bjDAZDSgur1TpFRx5BQM2nuTvmR7+dEROlqnScuFhlZItFoWh0++34bmXXleOSxBuwzwUM/K975jJ6DJ9Ldr4fWxOM50fOmiNOnBR54VHMeujVFzMLVFM1DCIu6yJgtIKvPtlNjzfScAza2LgLM5TydSJSr7boAXXa00JfnJmCHqPm8a086ljZYRiqP+oCXh62hol02ZWBDb+2EJrtEVdn3B7+xQ+SclRLNSQhRqjEd/erUHo+Ty4bT2lHJQaUo3Rdg1rZ6RvMHVEqaQmuk9cjP4+v/yeHRcPbzI2D+3Yc9RacgD/tZYGWqM16hTq8RVRCPziMrILy1HLoq5lMBJQ5s0yvHLgPJzld2RMH0Q9kDog2vodRLfJQRjsOeb7uulHVrpOW61SzJyedljRGq1BB1vys/eGeOyIz8bN0krWhKSUUdWI1ES31cdZPw4GYYEaVaLUBCCNU4JwHe6p0q3nC68oxpwChJWmBkI2nAMjMGJHIg4m32CvqMK9WpNi42RmIYYypZz82MyEEd3zdiEORqD1/EOqLvqN8IWLiysGug5HD6+X0H7aes5un5M5O2OM1mgNpssjnJ+mfnAOKVQqkdh7LPCMgjLMOXgeHZdzAzUkap5tMCR1Iplin6CX72x4uHvC3csHnuMnYuDcYDweEGpmTvtsHbRGa8wPU3PUxuOZyC+pYCAmpVgfJt1A93WcZlVKaZ5zFHRUxGL4wm14bUEglgatwM73/o690SlwociIH9rnLNAarcEFfrgyGttis1R9iNxKB4/JKESfDewXio1G1IYtKNetl0Ri9r4ERCemIisjDTm5eYhML8SQrYlweoPpq3vOAq3RGgykMwN5Jz4Lt8oq2cVNqKquxj9T89CLHb5+uXUQdaPPa4cuI7PIAHx3H7W1tQi7lI8Bf+GB+T0ERp5cdRy72PCKy6vwL/aNG0XlWHMsXU2+KrV0zzkKBvLosii8FZaOHI45X3Psqa424OjFPAzcxEAeCiMMRKRXGJEGWMgU28pU62KRXd1zjkAYZSBPrTqhak/S9y7Tt4K1eOhcDvptJPNNZoQLPLX6BPadvo5SkV5uUlFVjQ+4YRfa7W7QULDWWi+JwOrwdFWLah8GEno+9yGlFk+8I1NoFTfIvlWmTksm3aSrRegrw6GoSVNrhL1Kmu6gzV+SgVw1hMoekgFrIzLQ7Y88MHvMa43WIOVyUuP+egbR6QWqEX4n6cVNAkK/UkNikwPhYXTkCPR2TCZKKqpUr5Jgzly9hZm8eT4mvUpERfesBVqjNWRg5GfPkDjsPnUV5Uwrod1oMOIsWXlxz1nzb+xtVB+YUm0p375cJz2/RDFx22TiBFGNwxfyMG5XknqhYfGjXmiN1pBmR4hy+R26iHM3ihUrEoyJwYREX+EdhaxIII4MjWpd/pZsDNzElKKcV/KQJHVra4xKIfclXUcvzngqCHvrao224KYtF/NmuD4OW2Ky1CbSFGt5cmn5pZTiDPRlc2yl7iEMxt7pCeS+wuA7c2pYHpampgUJQlLqa35eyClWI1CHILLRkBFIa9SBG7cjxXM4sssmcmrCjIH3dVGZIxfzMZOXLrXx/LpgLKcun7KG+ttsk/mtC1mWK8AZpugdricHY2DvkLTacTJbZUGDWdYa64HMQs8Fx8GfRS6qJUHIOC/FL5+pTLtNJzIxeudpNTGrgMRxXpGV6oiE/uGIevsiLy52nrzKQylBJScF88sLI2+blTiQnINR7yaZmZBAbPzQQmusD+LUm+H48bpYJccZTCsTWTEyoNt0RMYXuauIuv1ufyp8d5/Fqx9fgOf2RLhtOaWuAiIObx1NQ9yVQnVVZrGpehN2s3hJ++jsDbzA59qzy5uZbAAbAq3xQWDKtFsWCfftCZTLLDXOS5rJaYpj8l0cu0L7pbwSFLGe5JXRF1SgeN5f5IapOjdhSU+pN7HLJc2bbKrRx5KaOh900BrtgZu0DoxU4/Uman96fplSHHm/ZWSxmvPdHJh8lxQUyP8siidBCHu5xRVIyCpU6/Rmg1WCIXs4EoRAa7QH2YTByNuSrlSdycz3/UyJZF68LuUWo5RN7R5rRpwWyOmL82bWDKjiFTmvuBzJ14p4z7mCX7x3Br3/HI9WUlci444GIdAaGwqhn1NpGzowmOPFdHbhgNBL2HwiC1FpBSooEYB01tJFFrW8qPgqrxQHUnIQdCRNvS+WMUcJgzjfEHWqD1qjoxBloQi04SjzNAfJnwTHUnVOK0cn70vGb/5xDtN5VRapFYl23ZKgpFV1bHne0XrQQWtsLORExSmBSKctRILlU9JHOS/QrNMYaI1NgUoRQZ2zOjzMACzQGpsjtMbmCK2xOUJrbI7QGpsdjuHf3S99jHLLRqMAAAAASUVORK5CYII="
style="max-width: inherit; cursor: pointer; color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size:
medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing:
normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px;
-webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness:
initial; text-decoration-style: initial; text-decoration-color: initial; width: 25px; height: 22.1154px;" width="50"
height="45">. Otherwise, it will have this
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAeCAYAAABNChwpAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAEnQAABJ0Ad5mH3gAAABdaVRYdFNuaXBNZXRhZGF0YQAAAAAAeyJjbGlwUG9pbnRzIjpbeyJ4IjowLCJ5IjowfSx7IngiOjMyLCJ5IjowfSx7IngiOjMyLCJ5IjozMH0seyJ4IjowLCJ5IjozMH1dfdp0gPkAAAQ1SURBVFhHxZdbbJNlGMf/PR/Wdls7mOtwEGQDHAREVEaMCSYmHiAhhDvFW+IFmnjlBSbeCndwzQ1cwI3BSBZJuDABBTXqPACTo8DObh1du3Zt13Y8/3fft77d15V1HPZLmvX7+h3+7/M+z/N/ZsPn3TNYRuzG32Vj2QUsegvsNhuiITc2v1CHRq8T7REvtq8KobO5Dh6HHTdjU/jxXhw/3Injz6FJjE/ljTur81gB8l6EfS68s64RO1eH8F5HGCvqXMavJQrylERmGoOJHG6PZ3DlQQJnr41hZDJnXFGZqgIcdhteXunHwdda5OX1WBv2Gr9Up1CcwYAIOXt9DN/1xtAzmIScqogDXR9+ZXwvQ96NrrYQvnx7Dd7taESTtupYehp3ZZVctdthg5MXa3C76mWbOkV8c8CtojCYzGGmgogFI/BqaxCf7WzF++vDcy+4MZrG9f/T+GNwEsevDKhzJOhxIOJ3qc/ejREckvtM8rL0n+4ncPjCf7g6krKIqBgBqv7olWbs64zA53Koc7/0JXHkUj+OXurDr/1Jdc4kJ6GIZ/IYklUyMn9JEtokCh1Nvrnk9Tnt6hlT00XjrlkqCjggL/9YPitFCOm+MY6vL/apDH8cEyLk6khahbw94seqeo/KpYDHiWE5xwjqQbD0gbDPqfa+VVSbnOsdw8+S1bXA64/JNnHbCBOYVcRo6FgEbG0JYJ1cTNXk+OUBKanykC+W7n9jasUmLGH2ER2LgF0vNZSp/FbK6H48YxzVDhOWVUPYP9jEdCwC3lzTgJC3vOSeBFYLk9OEHVTHIqAj4oOU9hxPKoCwPE3YvnUsArKFUplk80UkswXjaOkwsU3oHToWAdekWZgs1lCqwSblkR5gQuPSsQj4rb9Ubk87/GzddE0di4BbsVLGtwTdONRVaqtLQRdAt6Rl61gEPJRONpqaXTlv3hYNqO9LYXWDV3mDCa16fje1CPhnOIXvb44bR1B2/MGG0kNqoastOGdMtGjOCRxWdCwCqPKyuBdNhaxf4censg07pD3XAq/fs7HJOIKaDzikzE9sixnRKNg42I43yOrZkmkojTIV0fuT2aIynGowYl+81YZdaxvUcTKbx+m/R3GqZwSpXHlZV3TDjNR/30RW7HTWzWiptNbd8uC8pHJCegNLqyDmTis2YcJ+8kYU+zc1YYt4CuE8cOF2HCfl5bfmVQBZcCDhyne8GMTB16Oyknqp5/IebsJmxbCyZFk1etbnpKnRyI5cfKDCX2ksW3Ak4+QykMiKEWWRliEiGvQg4HaoaOhwWmKz4ezgN4YXwhw6L8nMAaZnKFX7TEh4D/OhVyx1UvZuSlbLbHZJLridFDN7nQ5L+BuZhs/Inp/4fVg56UIvJ4v+v4Cwp3NeoGXTNWlc9A62b3ZQNjH2EZYyq6lYaQqdR00CngWWPvB8AR4Buu7Olh/yrToAAAAASUVORK5CYII="
style="width: 22px; height: 20.9px;">&nbsp;aspect.</p>
<p>- Resetting the program will reset all your variables and previously performed analysis.</p>
<p><br></p>
<p><strong>Possible Problems and Suggested Fixes</strong></p>
<p>- A rare bug that occurs when running the program is the de-formatting of the pages that can happen when you go
back and forth between different pages causing artifacts of different pages or empty spaces before the pages to show up.
We currently believe this is an issue with the panel package we are basing this software on. This can be mostly solved
by refreshing the page in the browser to reset the layout of the page. However, <strong>do not use&nbsp;</strong>this
trick on the <strong>Class Intersection Analysis page</strong> since this has a small chance to make the sidebar to
navigate the software disappear, rendering it unoperable and needing to be closed and opened again.</p>
<p>- Another issue of the same type is the <strong>reset floatpanel</strong> not appearing when pressing it the 2nd time. It
is once again a problem fixed by refreshing.</p>
<p>- Do not mash the buttons many times, it will buffer the clicks. See first on the upper right if the program is running or not.</p>
<p>- Other known issues have smaller consequences and affect specific pages. When they exist, a little note in bold will
indicate the problem at the beginning of the corresponding page.</p>
<p><br></p>
<p><strong>Performance of the Different Parts of the Program</strong></p>
<p>- The program generally runs smoothly and quickly with low latency. The exceptions are the very first page
when data reading which can slow down the program for a few seconds as well as some parts when certain analysis such as
PLS-DA or Random Forest finish and the page layout is updated with the results.</p>
<p>- The possible slow downs and time of analysis is also dependent on the size of the dataset. The bigger the dataset to
be analyzed, the slower the analysis is and vice-versa.</p>
<p><br></p>
<p>We hope you find this software useful.</p>
<p><br></p>
<p>With our best regards,</p>
<p>The FT-ICR-SMS Laboratory NEMeSIS Development Team</p>'''


# Data Reading Description HTML in string format

data_reading_opening_string ='''<strong>Input your Excel or csv file.</strong>
<br>
This process can take a while for larger datasets.
<br>
The inputted file can either have samples in the columns and features on the rows or vice-versa. In any case, the first
column/row should have the feature <em>m/z</em> peaks, representing the metabolic features either their "neutral mass"
or <em>m/z</em> values obtained in either positive or negative ionization mode (choose accordingly based on your data).
If it is the former, a Neutral Mass column/row will be generated with those values; if it is the latter, a Probable
<em>m/z</em> row/column will be generated with those values. Albeit not necessary, it will allow to perform annotation
later on in the software. <strong>Avoid the existence of a column/row named "Neutral Mass" or 'Probable <em>m/z</em>' on
your data.</strong> This will only happen if the <em>m/z</em> peaks can be interpreted as a float (number). Otherwise,
a notification will inform that this column could not be added. If it could not be created in this way, you will
not be able to perform annotation (database annotation or formula assignment) using this software.
<br>
<br>
In general, the file will be transformed to have detected features in the rows and sample data and metadata in the columns.
You will be asked later to identify the metadata columns to separate them from the sample ones.
<br>
<br>
Currently, after confirming your dataset and moving to the next page, you will not be able to change it unless you restart
or reset the program.
'''


# Parameter Loading Description HTML in string format

load_parameters_string = """Here, a previously saved file with chosen parameters can be loaded into the interface. This file
comes from saving your parameters after pre-treatment (for only pre-processing and pre-treatment related parameters) or
in the report generation steps of the interface (for pre-processing, pre-treatment and data analysis related parameters).
<br>
<br>
This file <strong>does not include every single parameter</strong>. It does not include parameters that are heavily
dependent on the dataset being analysed such as the control and test class in univariate analysis as an example. It does
include some parameters that are <strong>still very dependent on the dataset such as the number of components and folds in
cross-validation for PLS-DA model building which should be tuned based on the dataset itself</strong>.
<br>
<br>
We include this possiblity in order for the user to setup a default set of parameters they like to speed analysis. However,
since every dataset is unique, we encourage the tuning of parameters for each dataset used.
<br>
<br>
A full list of parameters loaded in can be viewed by opening the corresponding json file."""



# Data Filtering Opening Description HTML in string format
data_filtering_opening_string="""Data Filtering can be performed here in 4 successive stages, each with their purpose and
objective. Each stage can be skipped as well and we even recommend to skip some of them depending on the type of data you
are utilizing.
<br>
<br>
The stages are:
<br>
<br>
<strong>1) Filtering based on the number of samples each feature appears in.</strong> This will only include features that appear
in a determined number of either the total samples of the dataset or of at least the samples of one class. (Filter experimental
artifacts, for example).
<br>
<br>
<strong>2) Filtering based on intensity values.</strong> This will filter metabolic features based on their intensity values as
calculated by the mean or median over the samples they appear in, removing lower intensity features. A hard threshold with a set
intensity value is used or a percentage based threshold where the percent lowest intensity of features are removed.
<br>
<br>
<strong>3) Filtering based on the feature variance of Quality Control samples.</strong> Only available if you have at least three
QC samples as selected in the previous page. This will eliminate features that have high variance in the quality control samples
(where they should be equal) as estimated by relative standard deviation (standard deviation / mean) above a determined threshold,
since that points that they are not reproducible. Features that do not appear in the QC samples are assumed to have 0 variance.
<br>
<br>
<strong>4) Filtering based on the feature variance of analysed samples.</strong> This filtering is not recommended to use if you
are looking for exclusive metabolic features in the classes of your datasets or if you aim to use feature occurrence data. It
will filter out the lowest variance features across the samples since their intensity patterns would not be informative.
"""



# Formula Assignment Opening Description HTML in string format
formula_assignment_opening_string = """Here Formula Assignment is made according to our ScoreForm algorithm. In-depth
description of the criteria and steps of ScoreForm are detailed in the **preprint**. It uses its default formula
database unless it is changed with the FormulaDatabaseCreator.ipynb notebook (if changed make sure their name follows
the convention 'formulas_improved_dict{lowest molecular mass allowed in file}.csv', for example 'formulas_improved_dict250.csv').
It is geared towards assignment up to 1250 Da. To perform Formula Assignment, the FormulaDatabaseCreator.ipynb must be ran
at least once before.
<br>
<br>
Parameters used that are shared with data annotation have to be changed in the data annotation page (even if it is not
performed) to make sure the same parameters are used in both procedures. Changeable parameters are formula assignment
specific.
<br>
<br>
When choosing the name of the formula column, make sure it is not the same name of a pre-existing column."""



# Data Multiple Annotation De-Duplication Description HTML in string format

annotation_deduplication_opening_string = '''<p>Due to the proximity of mass values, the same compound (or formula) can be
annotated to the different metabolic features.</p>
<p>Since one (adduct of one) compound should result in only one (monoisotopic) mass, this section <strong>can merge
metabolic features that share compound annotations into one single metabolic feature</strong>. For example, there is no
biological reason for multiple peaks being associated with a compound. Most of the time, there is one <em>&apos;main&apos;
peak</em> with much higher intensities across the samples, although this is not always observed.</p>
<p>Multiple situations can arise with different possible solutions, specifically a situation that warrants individual
attention that can arise when multiple annotations are performed on the data (to be avoided). When finding different
metabolic features with the same annotation in a database used, features are <strong>merged by keeping the highest
intensity value in each sample if representing the same adduct and summed if they correspond to different adducts</strong>
of the features with the same annotations. Thus, there is a total of 4 possible situations:</p>
<p>- <strong>Situation 1 (Overwrite)</strong>: Multiple metabolic features have the same annotation for all annotation
databases with the highest intensity values coming all from one metabolic feature that becomes the <em>de facto</em>
metabolic feature with others being erased.</p>
<p>- <strong>Situation 2&nbsp;(Merge Same Adducts)</strong>: Multiple metabolic features have the same annotation for all
annotation databases (and no other database has different annotations for those features) with the highest intensity coming
from at least two different metabolic features and <strong>ALL</strong> features come from the same adduct. For each sample,
the highest intensity is kept. &apos;Metabolite Label&apos; and the &apos;Mass&apos; columns become the weighted average (based
on the average intensity of the features) of all the features with the same annotation.</p>
<p>&nbsp;- <strong>Situation 3&nbsp;(Merge Different Adducts)</strong>: Identical to Situation 2 but there is at least one
metabolic feature that comes from a different adduct. For each sample, the highest intensity of metabolic features from the
same adduct are kept. Then, the intensities of metabolic features of different adducts are summed.</p>
<p>- <strong>Situation 4&nbsp;(Contradictions - Possible Problem Situation)</strong>: Multiple metabolic features have the
same annotation for one annotation database and different for another. Most of the time there is no issue. For example, imagine
a case where we have annotated HMDB compounds and ScoreForm assigned formulas. <strong>Scenario 1:</strong> We find
that HMDB puts two different compounds for 2 metabolic features and ScoreForm assigns the same formula. Here, it is fair to treat
them as different features and this is the default behaviour. <strong>Scenario 2:</strong> However, rarely, there can
be a case where there are <strong>more than 2 metabolic features</strong> with the same compound annotated in HMDB. This
should <strong>be quite rare and it is from where the problem arises</strong>. A scenario can arise where HMDB puts the same
compound for 4 features and ScoreForm assigns to one of them one formula, to a second one a different formula and the last two
no formula at all. What is the correct course of action? &nbsp;Perhaps merging the two peaks with no annotation by ScoreForm,
maybe merging those two peaks with one of those annotated by ScoreForm since they would be normally merged if not for the
existence of two different ScoreForm annotations. Hence, the problem. They should be seen on a <strong>case-by-case
basis&nbsp;</strong> and are thus not merged.</p>
<p><br></p>
<p><h3>Initial report of the observable multiple annotations for each annotation (including formula
annotation) performed</h3></p>'''


# Two strings for Parameter Saving Description HTML in string format

save_parameters_pretreatment_string = """Here, a json file can be created saving the parameters currently used for the
data pre-processing and pre-treatments steps that can be re-utilized in most analysis (dataset very specific parameters
such as metadata columns will not be saved). When loading in, there are some parameters that might not be applicable such
as the feature used in Normalization by a Reference Feature or the minimum number of samples used in data filtering. For
these cases, a check will be made to see if they can be used and if not, the default value will remain."""

save_parameters_dataanalysis_string = """Furthermore, relevant parameters for the different analysis steps will also be
saved. These include figure related parameters and methodology related parameters, but not dataset class related such as
the control and test class in univariate analysis."""


# PLS-DA Description HTML in string format

plsda_opening_string = """Here, <strong>Partial Least Squares - Discriminant Analysis</strong> models can be built.
<br>
<br>
As the number of components (Latent Variables) to use in a model such as this one is of essential importance, first we will
start with an optimization of this number. This is performed by fitting a PLS model in your data from a chosen initial
number of starting components to a chosen final number of components. The <strong>Q<sup>2</sup> and R<sup>2</sup></strong>
scores are then estimated by stratified cross-validation (number of folds <strong>K</strong> is also chosen by the user).
The chosen number of components should be the number <strong>where Q<sup>2</sup></strong> stops increasing. A
figure is provided to evaluate this. The maximum number of components chosen cannot be higher than the number of samples
that will train a model minus 1. For example, if you have 15 samples and a 3-fold cross-validation each fold will have 5
samples. A training set will be comprised of two of those folds thus it will have 10 samples, thus the number of components
cannot be higher than 9. Another example if you have 22 samples and 5 folds, the folds will have 4/4/4/5/5 samples each. A
training set will have four of these folds and the minimum sum of them is 4+4+4+5-1=16, thus it cannot be higher than 16.
<br>
<br>
- <strong>Q<sup>2</sup></strong> - Mean squared error of PLS-DA predictions based on the test samples, thus it is ideal
to test if the model will overfit. This will increase until a certain number of components that should be chosen. Then
it usually stabilizes but from a certain point it might start to decrease which would mean the model is overfitting.
<br>
- <strong>R<sup>2</sup></strong> - Mean squared error of PLS-DA predictions based on the training samples used to train
the model (it will be higher than <strong>Q<sup>2</sup></strong> but it should not be used to choose the number of
components. This metric always increases with more components (estimation errors might lead to temporary decreases)
which means it will overfit the model eventually.
<br>
<br>
After this optimization, a PLS model can be fitted with the chosen parameters and number of components. Model performance
can be estimated by accuracy, recall, precision and F1-score (the latter 3 weighted by class) using stratified fold
cross-validation. This cross-validation will also be used to estimate feature importance - variable importance in projection
(<strong>VIP</strong> scores). These are calculated based on Keiron Teilo O'Shea provided code in <a
href="https://www.researchgate.net/post/How-can-I-compute-Variable-Importance-in-Projection-VIP-in-Partial-Least-Squares-PLS"
target="_blank"
rel="nofollow">https://www.researchgate.net/post/How-can-I-compute-Variable-Importance-in-Projection-VIP-in-Partial-Least-Squares-PLS</a>.
This calculation can be slow when you have a high number of features. Importances estimated are then averaged across
the different folds. This can be iterated multiple times using randomized cross-validation folds. Model performance can
also be estimated with a Receiver Operating Characteristic (ROC) curve when only 2 classes are present.
<br>
<br>
Furthermore, a projection of the samples based on the PLS regression model using all available experimental samples on the
chosen Latent Variables akin to the PCA projections in the unsupervised analysis page is provided. Finally, Permutation
Testing (<strong>slow</strong>) can be performed to observe if the model and results obtained are significative.
"""


# Random Forest Description HTML in string format

rf_opening_string = """Here, <strong>Random Forest (RF)</strong> models can be built.
<br>
<br>
An optimization of the <strong>number of trees</strong> to use to build the Random Forest models is available. This will
allow the user to choose an optimal number of trees by observing the estimated accuracy of Random Forest models (by
stratified cross validation). The usual pattern is that the accuracy will increase until a certain number of trees and will
then fluctuate around that accuracy. Since RF are somewhat resistant to overfitting, the value chosen will be one where the
accuracy has stop increasing but not too high as to make the model training too slow.
<br>
<br>
After this optimization, a RF model can be fitted with the chosen parameters and number of trees. Model performance
can be estimated by accuracy, recall, precision and F1-score (the latter 3 weighted by class) using stratified fold
cross-validation. This cross-validation will also be used to estimate feature importance - <strong>Gini Importance</strong>.
Importances estimated are then averaged across the different folds. This can be iterated multiple times using randomized
cross-validation folds. Model performance can also be estimated with a Receiver Operating Characteristic (ROC) curve when
only 2 classes are present.
<br>
<br>
Finally, Permutation Testing (<strong>slow</strong>) can be performed to observe if the model and results obtained are
significative.
"""


# Permutation Test Description HTML in string format

permutation_test_description = '''Permutation tests permutate  the class labels of the samples, that is, all classes will
be randomized while maintaining the same number of samples per class and classes. For each permutation, model performance is
assessed by stratified cross-validation exactly as the previous sections. These performances are then compared with the
performance of a single iteration of a model built based on the non-permuted model that appears with a red line. Thus,
this red line might not be exactly the same value as the average model performance obtained in the supervised model (PLS-DA
or Random Forest) fitting section.
<br>
<br>
Thus, this is a test to observe model performance significancy, that is, if it is better than a random model. If it is,
then the remaining results from the important features give meaningful information.
<br>
<br>
<strong>Note: Permutation tests are slow to be computed.</strong>
<br>
<br>
<em>P</em>-value is calculated using the following equation:'''


# ROC Curve Description HTML in string format

ROC_curve_description = '''
<strong>ROC Curves</strong> are types of representation of the performance of supervised model that plot the True Positive
Rate by the False Positive Rate. Hence, one class has to be considered as the <strong><em>positive</em></strong> class.
Thus, this methodology, although possible in multiclass cases, is more suited for when you have a dataset with 2 classes.
When you have more than 2 classes, a ROC curve will be computed for each class considered as the positive by fitting a
"1vsAll" model, that is, the class of each sample is either the current positive class or "Other". Thus, there will likely
be many more negative samples than positive samples in each case, greatly increasing the importance of the latter for the
curve of each class.
<br>
<br>
The model performance is determined by the <strong>Area Under the Curve</strong> or (<strong>AUC</strong>). When it is 1,
the model is perfect for classifying your data; when it is 0.5, the model is no better than a random coin toss.
<br>
<br>
Number of Iterations indicates how many times ROC curve results are computed with random k-fold stratified cross-validation.
This can help to have a more detailed ROC Curve especially when your dataset has a low number of samples.'''


# 1v1 Univariate Analysis Page Opening HTML Description in string format

univ_opening_string = '''In this section, both 1v1 Univariate Analysis and Fold-Change analysis are performed.
<br>
<br>
Here, a chosen test class and a chosen control class from the classes in the dataset are put in opposition to detect
metabolites with significant differences between the classes.
<br>
<br>
The Fold change is calculated in a dataset with missing values imputed and normalized after. <strong>This means that with
our very high number of missing values in FT-ICR-MS data, it affects the calculation of the fold change a lot. Thus, take
this fold change values with a grain (or multiple grains that are actually more like rocks than grains) of salt.</strong>
It is thus more suited to when there are a low number of missing values in your data since calculation of the fold change
between the classes is severely affected with missing values since it is calculated after missing values imputation and
normalization.
<br>
<br>
Choose between the parametric <strong>t-test</strong> and non-parametric <strong>Mann-Whitney test</strong>.
<br>
<br>
<strong>Warning</strong>: If you have more than 2 classes, a control class and a test class can be chosen to perform the analysis.
<br>
<br>
Finally, when you have more than 2 classes, each unsupervised analysis will select the samples respective to the 2 classes,
data filtering (Filter 1 only) and pre-treatment will be performed the same way it was done on the dataset before performing
unsupervised analysis (starting from the dataset <strong>after data filtering, annotation and de-duplication</strong>)*. In
case of filtering based on the total number of samples a feature appears in, the number used on the full dataset is converted
to percentage-based and then a new number is calculated based on the number of samples of the control and test class, which is
then rounded up.
<br>
<br>
* - We cannot start from the completely full dataset since we have no annotations associated with it despite the fact that,
when using total number of samples based data filtering, there could be some fatures that were initially filtered out that
could remain in the dataset. E.g. if you choose a minimum of 4 samples to appear in a 15-sample 5-class dataset. When
performing univariate analysis on 2 classes that have a total of 6 samples, the minimum 5 samples would be transformed to a
minimum of 2 samples in the 6-sample subset (4 out of 15 represents 26.67% of samples, which is 1.6 out of 6 samples rounded
up to 2). There could be features that appear in 2 of these 6 samples but that did not appear in 4 of the original 15 samples
and were thus previously removed from the dataset. These features <strong>stay removed</strong>.
'''


# Multiclass Univariate Analysis Page Opening HTML Description in string format

multiclass_univ_opening_string = '''In this section, Multiclass Univariate Analysis is performed.
<br>
<br>
Here, all classes are tested simultaneously. As such, no fold-change can be calculated or Volcano Plot
drawn, but the results show significant metabolites for all classes with the clustermap showing their
groupings and intensity pattern trends.
<br>
<br>
Choose between the parametric <strong>ANOVA</strong> and non-parametric <strong>Kruskal-Wallis test</strong>.
<br>
<br>
<strong>Warning</strong>: These tests can only be performed if you have more than 2 classes.
<br>
<br>
The resulting table can be downloaded.
'''



# Van Krevelen Description HTML in string format

vk_opening_string = '''<strong>Van Krevelen Plot section</strong>
<br>
<br>
This section plots a Van Krevelen Plot for each of the classes under analysis. This is made by only considering metabolites
(features) that appear at least in one sample of said class.
<br>
Only metabolites with assigned formulas are considered. You can choose which formula annotation you want to use whether it
is a previous annotation performed, an annotation performed in this software, or even multiple different annotations
(<strong>Note: you HAVE to select at least one annotation</strong>). If multiple formulas can be assigned to a metabolite
whether within the same database annotation or different, they are both considered and plotted in the Van Krevelen.
<br>
<br>
You can also decide to colour the dots and make them different sized based on their average intensity in the corresponding
class using either a rank-based (more linear) approach or the logarithm of the averages intensity (less linear). You can
also choose a "midpoint", where the dots with avg. intensity below the midpoint % of the intensity of all features have a
more bluish colour, while those above have a redder colour.
'''


# Kendrick Mass Defect Plot Description HTML in string format

kmd_opening_string = '''<strong>Kendrick Mass Defect Plot section</strong>
<br>
<br>
This section plots a Kendrick Mass Defect Plot for each of the classes under analysis. This is plotted based on Neutral
Mass column selected.
<br>
Kendrick Nominal Mass and Mass Defect can be calculated in two ways:
<br>
- <strong>'Up'</strong>: by rounding the masses up, thus the Mass Defects will vary between 0 and 1.
<br>
- <strong>'Nearest'</strong>: by rounding the masses to the nearest integer, thus the Mass Defects will vary between -0.5 and 0.5.
<br>
<br>
You can decide to colour the dots based on the chemical composition series they belong to ('CHO', 'CHON', 'CHOP', 'CHONSP',
etc.) by selecting the formula columns of previous annotations, of the annotations performed in the software or a
combination of them. When, for one metabolic feature, there are multiple candidate formulas and they do not belong to the same
chemical composition series, they get assigned as <strong>'Ambiguous'</strong>. If <strong>None</strong> is selected, then the
points are not coloured by chemical composition series.
'''


# Chemical Composition Series Plot Description HTML in string format

ccs_opening_string = '''<strong>Chemical Composition Series section</strong>
<br>
<br>
This section plots a Chemical Composition Series bar plot indicating how many formulas of different chemical series ('CHO',
'CHON', 'CHOP', 'CHONSP', etc.) were assigned to samples of each class.
<br>
<br>
You can decide if you want to use formula annotations made previously, made in this software or a combination. <strong>If
multiple formulas</strong> can be assigned to the same metabolic feature whether within the same formula annotation made or
between different annotations, <strong>each one will be counted in this plot</strong>. That is, if a peak in a class has 3
possible formulas, 2 belonging to the 'CHO' series and another to the 'CHOP' series; then 2 formulas will be added
to the 'CHO' series and 1 to the 'CHOP' series. Thus, we are considering that the 3 elemental formulas are represented by
that feature (probably an overestimation). To provide an idea of how extensive this effect is, a description is also provided
detailing how many formulas are being considered for each class and from how many different features (<em>m/z</em> peaks)
they came from.
'''


# HMDB Pathway Assignment Page Description HTML in string format

path_assign_opening_string = """This page allows you to assign or match known metabolic pathways to HMDB annotations only
(abbreviation should be HMDB). This is made based on a file created from the <strong>RAMP database</strong> (<a
href="https://rampdb.nih.gov/" target="_blank" rel="nofollow">https://rampdb.nih.gov/</a>) where 55872 HMDB compounds have at
least 1 corresponding pathway. The pathways shown are aggregated from multiple sources: HMDB, Reactome, WikiPathways and KEGG;
from where RAMP provides the information (<a href="https://academic.oup.com/bioinformatics/article/39/1/btac726/6827287"
target="_blank" rel="nofollow">https://academic.oup.com/bioinformatics/article/39/1/btac726/6827287</a>).
<br>
Thus, for this, there must be a column with the <strong>HMDB identifiers</strong> (e.g. 'HMDB0000001') from which the assignment
can be performed. The output gives the HMDB IDs and annotation names (if possible) as index in a DataFrame and the name and id of
the pathways that they belong to according to this built file. Your identifiers should have 7 numbers after "HMDB".
<br><br>
<strong>If no HMDB identifiers are present, you will not be able to perform this data analysis step</strong>."""


# KEGG Mapping Colour Section Description HTML in string format

kegg_opening_string = """This section allows to extract every KEGG compound ID annotated to the dataset and obtain a downloaded
file containing a list with only the KEGG identifiers and the colour associated with them (colours represent metabolite presence
in classes and can be chosen below). This file can then be used as input in the KEGG website at
<a href="https://www.genome.jp/kegg/mapper/color.html" target="_blank" rel="nofollow">https://www.genome.jp/kegg/mapper/color.html</a>.
<br><br>
This section is <strong>only enabled if the dataset only has 2 classes</strong>. KEGG colour mapping is not available for more than
2 classes.
<br><br>
KEGG identifier will only be mapped if a <strong>Matched KEGGs</strong> column exists. This column can be metadata before previously
included in the dataset or can be generated in this software during Data Annotation if <strong>at least one of the databases used for
annotation has a column with the name 'kegg' (with no capitalization) which will create the Matched KEGGs column.</strong>
"""


# HMDB Pathway Over-Representation Analysis Section Description HTML in string format

pathway_ora_opening_string = """
Pathway Over-Representation Analysis is performed based on HMDB compound pathway matching done before. The background set
can be either all the metabolites with HMDB identifiers with associated pathways in the <strong>RAMP
database</strong> (<a href="https://rampdb.nih.gov/" target="_blank" rel="nofollow">https://rampdb.nih.gov/</a>) (currently
not implemented) or <strong>restricted to the number of HMDB annotated metabolites in the dataset with associated
pathways</strong>. The background set to choose is critical. If the RAMP pathway database is used as background, most
metabolic pathways will appear as significant by 'common <em>p</em>-values' even after multiple testing correction and very
general pathways with more metabolites such as 'Metabolism' or 'Biochemical pathways: part I' or 'Transport of small molecules'
have a good chance of appearing as significant. Thus, pathways with multiple metabolites annotated that constitute a decent
part of the pathway should be looked at more carefully. If the <strong>HMDB annotated metabolites in the dataset with
associated pathways</strong> are used as the background, this conservative approach makes it more complicated for any pathway
to be considered significant but there is less bias per pathway. In this case, we recommend looking not only at the number of
significant and detected metabolites in the pathway but also to the overall number of metabolites in the pathway (in the
database), which is included in the output table.
<br>
<br>
Multiple peaks with the same annotation will all be considered as individual entries. Peaks that have multiple
annotations that could correspond to multiple metabolites in the same pathway will be counted as a single metabolite.
So, analysis of the relevant pathways should be performed. For each pathway (and each class), you may search for the
corresponding relevant metabolic features associated.
<br>
<br>
Select through which method is desired to select the significant metabolites and the threshold used in each case explained
next to them. The significant metabolites have to have associated at least 1 pathway. In each case it will only work if one
of the methodologies has been previously ran.
<br>
- <strong>RF Gini Importance</strong> - Threshold is the top % of ranks considered significant (e.g. 0.20 for 20%).
<br>
- <strong>PLS-DA Feat. Importance</strong> - Threshold is the top % of ranks considered significant if below 1 (e.g. 0.20
for 20%) or the threshold for importance if higher that should only be used for VIP Scores - <strong>Default</strong>.
<br>
- <strong>1v1 Univariate Analysis</strong> - Threshold is the maximum adjusted (for multiple test correction) <em>p</em>-value
from the univariate analysis (e.g. 0.05 for adjusted <em>p</em>-values under 0.05). The test class is the class chosen as test
during univariate analysis.
<br>
- <strong>Multiclass Univariate Analysis</strong> - Threshold is the maximum adjusted (for multiple test correction)
<em>p</em>-value from the univariate analysis (e.g. 0.05 for adjusted <em>p</em>-values under 0.05).
"""


# Binary Simplification (BinSim) Analysis Page Description HTML in string format

BinSim_opening_string = '''Here, unsupervised and supervised analysis can be performed by using <strong>BinSim treated
data</strong> instead of intensity treated data.
<br>
<br>
<strong>Binary Simplification or BinSim</strong> pre-treatment consists of transforming all <strong>intensity
values</strong> present in the data matrix to <strong>1</strong> (metabolic features present in the data) and all
<strong>missing values</strong> to <strong>0</strong> (metabolic features absent from the data). Thus, it is a strategy
focusing on <strong>feature occurrence</strong> instead of intensity patterns. This methodology is rather useful when
analysing data that is suspected to have considerable intensity variation between analysis (as FT-ICR-MS data tends to
have) as feature occurrence should be more robust and therefore reproducible. Furthermore, it is a complementing analysis
to the usual intensity-focused analysis offering a complementary perspective.
<br>
<br>
For more information on this methodology, check out our paper on it: <strong><a data-fr-linked="true"
href="https://www.mdpi.com/2218-1989/11/11/788">https://www.mdpi.com/2218-1989/11/11/788</a></strong>.
<br>
Citation: Traquete F, Luz J, Cordeiro C, Sousa Silva M, Ferreira AEN. Binary Simplification as an Effective Tool in
Metabolomics Data Analysis. <em>Metabolites</em>. 2021; 11(11):788. <a data-fr-linked="true"
href="https://doi.org/10.3390/metabo11110788">https://doi.org/10.3390/metabo11110788</a>.
<br>
<br>
In this section, BinSim treated data is used to perform <strong>PCA</strong>, <strong>HCA</strong> (unsupervised),
<strong>PLS-DA</strong> and <strong>Random Forest</strong> (supervised) statistical analysis.'''


# Graph-based Analysis Page Description HTML in string format

Graph_opening_string = '''Here, extreme-resolution direct-infusion metabolomics data can be used to generate either
Mass-Difference Networks (MDiNs) or Formula-Difference Networks (FDiNs). Moreover, in the later section, these generated
networks can be used to perform the sMDiN analysis according to the network analysis methods provided (more on that in
the appropriate section).
<br>
<br>
These MDiNs/FDiNs are built from the complete dataset using detected metabolites as nodes and establishing connections
between them if their mass (or formula) difference corresponds to one of the presented biochemical transformations used
to generate them. These transformations are called Mass-Difference based Building blocks or MDBs. As such, the <em>ab
initio</em> networks are similar to metabolic networks in a way that they represent the chemical diversity in the dataset.
However, their connections are not guaranteed to represent reactions that actually happen <em>in vivo</em>. On the other
hand, they do not require any sort of previous knowledge or information from metabolic pathways that are many times
incomplete (especially on less studied biological systems) to establish connections and can represent all sorts fo reactions
including non-enzymatic ones.
<br>
<br>
The MDBs are the list of chemical transformations that are used to build Mass-Difference Networks. This list can be inputted
to the software to use a set of transformations desired for the dataset in question. Moreover, a default list `MDB_list.txt`
with 15 common biochemical transformation can be used as default. This file can also be used as a guide for the formatting
required to create an MDB list. In short, it should have <strong>4 columns without headers</strong>. The first one
corresponds to the absolute change in the compound that the chemical transformation causes (for example, methylation is CH2).
<strong>See `MDB_list.txt` how to represent these changes</strong>, especially when there is a negative change of some atoms.
In these cases, such as O(N-H-), make sure the atoms where there is a negative change are between brackets and that thye have
a '-' after their symbol. The second column has the name of the Mass-Difference-based building Block (MDB), that is of the
chemical transformation, the third column the mass difference associated to the column (for methylation, it would be
14.015650064399999 for example) and the fourth column should say true. Each column should be tab separated.
<br>
<br>
To build the networks, there are 4 possible algorithms available:
<br>
<br>
- <strong>Simple MDiN</strong>: This is the simplest and fastest type of MDiN. It only requires a list of mass values (and MDBs)
and creates connections (edges) in the network if the mass difference between metabolic features (nodes) corresponds to one of
the MDBs used.
<br>
- <strong>Univocal MDiN</strong>: It also only requires a list of mass values (and MDBs). It starts by building a simple MDiN but
then it trims this simple MDiN by seeing cases where one node has multiple edges representing the same chemical transformation in
the same 'direction' (that is addition or subtraction of a chemical group) and only keeping the one who has a lower associated
error. This is based on the idea that each metabolite feature should have a unique formula, since direct infusion mass
spectrometry would indicate that for each unique formula, you should only have one metabolic feature. If there are multiples of
the same 'edge' MDB from the same node, this principle would be violated.
<br>
- <strong>Mass Formula Propagation (Mass Form. Prop.) MDiN</strong>: It requires a list of mass values and a list of reliably (or
more reliably) identificated metabolic features (for example, formulas from annotated metabolites). It builds a Univocal MDiN but
uses it for formula propagation in the network starting from those reliably identificated metabolic features. It eliminates edges
to both avoid incoherencies and contradictions. The former by leading to very improbable formulas as defined by different
criterias such as the usual elemental ratios, Valency restrictions or maximum number of different elements. The latter by
eliminating edges where the propagation from two starting nodes leads to different formulas.
<br>
- <strong>Formula Difference Networks</strong>: This network is built from assigned formulas in the data instead of the mass
differences. It requires only a DataFrame or Series with the formula associated to each metabolic feature. It restricts the data
to only the formula assigned metabolites but also reduces possible errors in link creation since there is no deviation associated
with formula differences.
'''


# Report Generation Page Description HTML in string format

report_opening_string = '''This section can provide a <strong>report of the metabolomics dataset</strong> analysed here. It
includes the sections of statistical analysis chosen below.
<br>
<br>
The report will be generated based on the parameters you have chosen for the different analysis which were saved.
Furthermore, to perform the generation, <strong>all relevant figure and tables will also be downloaded exactly as they are
right now.</strong>
<br>
<br>
A folder with the chosen name will be created in your working directory. The report as well as all images
and tables present or mentioned in said report will be downloaded and stored in this folder. Once again, figures will be
shown <strong>exactly</strong> as they were left in the analysis. Please select a folder name that does not yet exist in
your working directory.
<br>
<br>
<strong>Note</strong>: An attempt is made to correctly describe the parameters that led to the results obtained in each
part but we cannot be sure that the figures shown are 100% represented by the parameters described since errors may arise
in parameter saving.
'''